libvuln/updates/manager.go

```diff
@@ -0,0 +1,256 @@
+package updates
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"runtime"
+	"strings"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/quay/claircore/internal/vulnstore"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/pkg/distlock/postgres"
+	"github.com/quay/claircore/updater"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	DefaultInterval = time.Duration(30 * time.Minute)
+)
+
+var (
+	DefaultBatchSize = runtime.NumCPU()
+)
+
+type Configs map[string]driver.ConfigUnmarshaler
+
+// Manager oversees the configuration and invocation of vulstore updaters.
+//
+// The Manager may be used in a one-shot fashion, configured
+// to run background jobs, or both.
+type Manager struct {
+	// provides run-time updater construction.
+	factories map[string]driver.UpdaterSetFactory
+	// max in-flight updaters.
+	batchSize int
+	// update interval used once Manager.Start is invoked, otherwise
+	// this field is not used.
+	interval time.Duration
+	// configs provided to updaters once constructed.
+	configs Configs
+	// provided to construct distributed locks.
+	// TODO(louis): this will be replaced by newest distlock implementation.
+	pool   *pgxpool.Pool
+	client *http.Client
+	store  vulnstore.Updater
+}
+
+// NewManager will return a manager ready to have its Start or Run methods called.
+func NewManager(ctx context.Context, store vulnstore.Updater, pool *pgxpool.Pool, outOfTree []driver.Updater, client *http.Client, opts ...ManagerOption) (*Manager, error) {
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	// the default Manager
+	m := &Manager{
+		store:     store,
+		pool:      pool,
+		factories: updater.Registered(),
+		batchSize: DefaultBatchSize,
+		interval:  DefaultInterval,
+		client:    client,
+	}
+
+	// out-of-tree updaters is a required dependency
+	// even tho it could be empty.
+	//
+	// this avoids a dependecy issue which would force ManagerOption(s)
+	// to be ran in a specific order.
```

Comment:
The original error scenario this was intended to avoid was: 

- user constructs manager placing WithConfigs option *before* OutOfTree option
- WithConfig option calls updater.Configure with default set of factories, excluding the synthentic "outOfTree" factory we create
- outOfTree factory set never gets configured. 

taking a look again tho, it appears we always use a static factory set for outOFtree, which will never be configurable. 

I guess this bears the question then, will we ever have out of tree factories which are configurable? If so, this should stay as is. While we do hard code it in our code, anyone who refactors or uses the manager in another place is susceptible to this issue if outOfTree is not a required dependency to the constructor.

Reply:
alright @hdonnay I removed the required outOfTree in favor of the options. 
If we decide that the synthetic outOfTree factory needs configuration at some point, it should become a dependency again to avoid the configuration ordering issue.